### DISCUSSION TITLE
INTEGER() == INTEGER() is False

### DISCUSSION THREAD

**Discussion by joarobles** [sqlalchemy/sqlalchemy#9195] [Usage Questions]
I don't seem to be able to compare SQLAlchemy data types:

```python
>>> from sqlalchemy import INTEGER
>>> INTEGER() == INTEGER()
False

>>> from sqlalchemy import VARCHAR
>>> VARCHAR() == VARCHAR()
False
```

This causes objects fetch by `Metadata` reflection not being able to be compared for equality, should this be considered a bug?

Also, what's the difference between `Integer` and `INTEGER`? When I declare my columns it's recommended to use `Integer` but the result of the metadata inspection would return an `INTEGER` instance for the `type` attribute.

**Comment by CaselIT** [comment]
hi

types don't have an equal implementation, since the actual sql type depends on the dialect.
It's better to compare the compiled form using the actual dialect in use.

```py
INTEGER().compile(engine.dialect) == INTEGER().compile(engine.dialect)
```

> Also, what's the difference between `Integer` and `INTEGER`? When I declare my columns it's recommended to use `Integer` but the result of the metadata inspection would return an `INTEGER` instance for the `type` attribute.

you can find the rationale at https://docs.sqlalchemy.org/en/20/core/type_basics.html
